### PR TITLE
#68/conversation card menu

### LINKED ISSUE #68: Replace x modal option with three dots 🍒
The dots should take you to a menu that offers the possibility to delete the conversation or to mark the entire conversation as unread (do not work on the latter as there is a separate issue for this)

The menu should look like this

Image

### PR DESCRIPTION
**Pull request by nichgalzin**
# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review on my code
- [x] I have made corresponding changes to the documentation (if any were needed)
- [x] My changes generate no new warnings or errors
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and previously existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# Description

**Relates #68 / Closes #68**
Include here a short description of what behaviour this Pull Request introduces.
Outline any new dependencies introduced.
- Add a meatball menu to each new conversation card
- Adds a drop down menu with a delete conversation option
- Display a confirmation modal to confirm deletions. 
- Makes sure conversations are properly deleted and cascading deletion when both user delete a conversation.

### Files changed
- app/styles/messaging-styles.css -- added overlay custom class
- components/DeleteConvoButton.tsx -- deleted as a generic component was used instead
- components/DeleteConvoModal.tsx -- deleted as implementation was refactored
- components/buttons/ButtonRounded.tsx -- added functionality to make generic component more reusable
- components/messaging/CardMenu.tsx -- created.
- components/messaging/ConversationCard.tsx -- Add menu and moved modal and functionality here.
- components/messaging/ConversationCardModal.tsx -- new component to handle delete warning modal
- components/messaging/ConversationsList.tsx -- Moved modal to down to conversation card
- components/messaging/MessageCard.tsx -- removed tick icon
- utils/messaging/deleteConversation.ts -- improved error handling
 
Include one to two lines on the necessary changes to each file.

### UI changes

Screenshot 2024-03-14 at 15 10 14

# Tests

Include a description of any tests added - if none have been added explain why.

No tests where add as still working out the bugs with our component testing implementation.

### CODE REVIEW

File: components/messaging/CardMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
+'use client';
+import { useState } from 'react';
+import MeatballIcon from '../icons/MeatballIcon';
+
+type CardMenuType = {
+  toggleModal: (event: React.MouseEvent<HTMLButtonElement>) => void;
+};
+
+const CardMenu: React.FC<CardMenuType> = ({ toggleModal }) => {
+  const [menuOpen, setMenuOpen] = useState(false);
```
Comment (camelPhonso) [inline components/messaging/CardMenu.tsx:10]: `isMenuOpen` just to keep a convention on naming our booleans 🦆

File: components/messaging/ConversationCard.tsx [hunk 2]
```diff
@@ -22,24 +23,28 @@ const ConversationCard: React.FC<ConversationCardProps> = ({
   joinedAt,
   itemName,
   imageSrc,
+  conversationId,
   clickHandler,
 }) => {
   return (
-    <button type='button' onClick={clickHandler}>
-      <div className='m-2 flex max-h-28 w-[400px] items-center gap-4 rounded-lg bg-gray-300 p-4 shadow-md hover:shadow-lg'>
+    <div tabIndex={0} aria-label='button' onClick={clickHandler}>
```
Comment (camelPhonso) [inline components/messaging/ConversationCard.tsx:30]: Smart 🧠 

I figure it might be a bit unconventional but can't think of any better way to solve that button issue. Nice touch that you added the aria-label!

File: components/messaging/ConversationCardModal.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import CardMenu from './CardMenu';
+import ButtonRounded from '../buttons/ButtonRounded';
+
+interface ModalProps {
+  conversationId?: number;
+  message: string;
+}
+
+const ConversationCardModal = ({ conversationId, message }: ModalProps) => {
+  const [modalOpen, setModalOpen] = useState(false);
```
Comment (camelPhonso) [inline components/messaging/ConversationCardModal.tsx:17]: `isModalOpen` 🦆

File: utils/messaging/deleteConversation.ts [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
+    const { data, error } = await supabase
+      .from('user_conversations')
+      .delete()
+      .match({ id: convIDtobeDeleted });
+
+    if (error) throw error;
+    console.log(data);
   } catch (error) {
-    console.log(error);
+    console.error('Error deleting conversation:', error);
```
Comment (camelPhonso) [inline utils/messaging/deleteConversation.ts:14]: Nice! Thanks for catching this one 🦅 (that's because you were eagle-eyed)

File: components/messaging/ConversationCardModal.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  const [isDisabled, setIsDisabled] = useState(false);
+
+  const toggleModalClickHandler = (
+    event: React.MouseEvent<HTMLButtonElement>
+  ) => {
+    event.stopPropagation();
+    setModalOpen((prevState: boolean) => !prevState);
+  };
+
+  const deleteConversationClickHandler = async (
```
Comment (camelPhonso) [inline components/messaging/ConversationCardModal.tsx:28]: I'm not as set on this one, but should we include the event on our Handler names or just the action performed?

`deleteConversationHandler` instead of `deleteConversationClickHandler` ?

The handlers could theoretically be re-used in other events and it makes the names a bit shorter and simpler without resorting to abbreviations.

File: components/messaging/ConversationCardModal.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import React from 'react';
+import { useState } from 'react';
+import '../../app/styles/messaging-styles.css';
+import deleteConversation from '@/utils/messaging/deleteConversation';
+
+//Components
+import CardMenu from './CardMenu';
+import ButtonRounded from '../buttons/ButtonRounded';
+
+interface ModalProps {
```
Comment (camelPhonso) [inline components/messaging/ConversationCardModal.tsx:11]: ahhhh an `interface` in the midst of our `type`s ...

can we stick to one?

File: components/messaging/CardMenu.tsx
Comment (camelPhonso) [inline components/messaging/CardMenu.tsx:1]: I think there's a refactor that can happen here. Though you let me know if it doesn't make sense:

Could this be named `EllipsisMenu.tsx` and take an extra argument for re-usability? 
(EllipsisMenu because I'd genuinely rather not have Meatballs... 🥬 )

With a few more props this can be widely re-usable across the app (beyond messaging).

```ts
type EllipsisMenuProps = {
  handleFirstModal: (event: React.MouseEvent<HTMLButtonElement>) => void;
  handleSecondModal:(event:React.MouseEvent<HTMLButtonElement>) => void;
  firstButtonMessage: string;
  secondButtonMessage: string;
};
```
You get the idea.

**Review by camelPhonso — CHANGES_REQUESTED** [review]
Cool stuff. 

I'm mostly picking on your naming of stuff but take a moment to review the suggestion on refactoring `CardMenu.tsx`

File: components/messaging/ConversationCardModal.tsx [hunk 7]
```diff
@@ -5,24 +5,24 @@ import '../../app/styles/messaging-styles.css';
 import deleteConversation from '@/utils/messaging/deleteConversation';
 
 //Components
-import CardMenu from './CardMenu';
+import CardMenu from '../menus/EllipsisMenu';
```
Comment (camelPhonso) [inline components/messaging/ConversationCardModal.tsx:8]: I'm confused how this is happening, but you are still importing CardMenu even though you've changed the export to be EllipsisMenu

File: components/messaging/ConversationCardModal.tsx [hunk 8]
```diff
@@ -49,8 +49,19 @@ const ConversationCardModal = ({ conversationId, message }: ModalProps) => {
...
+            buttonMessage: 'Delete Conversation',
+            clickHandler: toggleModalClickHandler,
+          },
+          {
+            buttonMessage: 'Mark Unread',
+            clickHandler: () => console.log('Log'),
+          },
+        ]}
+      />
+      {isModalOpen && (
```
Comment (camelPhonso) [inline components/messaging/ConversationCardModal.tsx:64]: This is way beyond what I had imagined! Looks great, thanks ✨

**Review by camelPhonso — APPROVED** [review]
Awesome rework, thanks for going through and refactoring the menu.

I'm approving but watch out for that weird naming thing on when you import the newly worked EllipsisMenu onto `ConversationCardModal.tsx`

I guess it's just renaming the default export though the syntax doesn't seem quite correct to me 😅 but in any case it's just making it a bit confusing (though it runs).
